traces/traces.go

```diff
@@ -0,0 +1,101 @@
+package traces
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	instrumentationPkg = "github.com/osquery/osquery-go"
+	osqueryGoVersion   = "0.0.0"
+)
+
+// OsqueryGoTracer provides a tracer with a standardized name and version.
+func OsqueryGoTracer() trace.Tracer {
+	return otel.Tracer(instrumentationPkg, trace.WithInstrumentationVersion(osqueryGoVersion))
+}
+
+// StartSpan is a wrapper around trace.Tracer.Start that simplifies passing in span attributes.
+// `keyVals` should be a list of pairs, where the first in the pair is a string representing
+// the attribute key and the second in the pair is the attribute value.
+// The caller is always responsible for ending the returned span.
+// Any spans requiring more specific configuration can be created manually via OsqueryGoTracer().Start.
+func StartSpan(ctx context.Context, spanName string, keyVals ...interface{}) (context.Context, trace.Span) {
+	opts := make([]trace.SpanStartOption, 0)
+
+	// Extract information about the caller to set some standard attributes (code.filepath,
+	// code.lineno, code.function) and to set more specific span and attribute names.
+	// runtime.Caller(0) would return information about `StartSpan` -- calling
+	// runtime.Caller(1) will return information about the function calling `StartSpan`.
+	programCounter, callerFile, callerLine, ok := runtime.Caller(1)
+	if ok {
+		opts = append(opts, trace.WithAttributes(
+			semconv.CodeFilepath(callerFile),
+			semconv.CodeLineNumber(callerLine),
+		))
+
+		// Extract the calling function name and use it to set code.function.
+		if f := runtime.FuncForPC(programCounter); f != nil {
+			opts = append(opts, trace.WithAttributes(semconv.CodeFunction(f.Name())))
+		}
+	}
+
+	opts = append(opts, trace.WithAttributes(buildAttributes(keyVals...)...))
+
+	return otel.Tracer(instrumentationPkg, trace.WithInstrumentationVersion(osqueryGoVersion)).Start(ctx, spanName, opts...)
+}
+
+// buildAttributes takes the given keyVals, expected to be pairs representing the key
+// and value of each attribute, and parses them appropriately, ensuring that the keys
+// have consistent and specific names. Pairs with invalid keys or values will be added
+// as string attributes.
+func buildAttributes(keyVals ...interface{}) []attribute.KeyValue {
```

Comment:
I'm not totally sure about the type switch in the SDK here. Might be a good enough balance though